lib/remap-functions/src/append.rs

```diff
@@ -0,0 +1,143 @@
+use remap::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Append;
+
+impl Function for Append {
+    fn identifier(&self) -> &'static str {
+        "append"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Array(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "items",
+                accepts: |v| matches!(v, Value::Array(_)),
+                required: true,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let items = arguments.required("items")?.boxed();
+
+        Ok(Box::new(AppendFn { value, items }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct AppendFn {
+    value: Box<dyn Expression>,
+    items: Box<dyn Expression>,
+}
+
+impl Expression for AppendFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let mut value = self.value.execute(state, object)?.try_array()?;
+        let mut items = self.items.execute(state, object)?.try_array()?;
+
+        value.append(&mut items);
+
+        Ok(value.into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::Kind;
+
+        let items_type = self
+            .items
+            .type_def(state)
+            .fallible_unless(Kind::Array)
+            .with_inner_type(self.items.type_def(state).inner_type_def);
+
+        self.value
+            .type_def(state)
+            .fallible_unless(Kind::Array)
+            .merge(items_type)
+            .with_constraint(Kind::Array)
+            .with_inner_type(
+                self.items
+                    .type_def(state)
+                    .merge(self.value.type_def(state))
+                    .inner_type_def,
+            )
```

Comment:
re: your question about `inner_type_def`.

I saw you that you tried to remove this entirely and just rely on `merge(item_type)` to also merge the inner type def. I assume that didn't work as expected, since you moved back to manually setting the inner type?

We definitely need to improve this. I admit that adding the `inner_type` was done fairly hastily to solve the issues not having this information caused in our type system.

Reply:
That tinkering was mostly rooted in ignorance 😄 It also works without the `with_inner_type` setters, so I suspect that the inner type reconciliation is already handled by `merge`, making `with_inner_type` redundant. Is that correct?